cmd/restore.go

```diff
@@ -0,0 +1,269 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+	"time"
+	"unsafe"
+
+	"github.com/spf13/cobra"
+	"github.com/tigrisdata/tigris-cli/client"
+	"github.com/tigrisdata/tigris-cli/util"
+	"github.com/tigrisdata/tigris-client-go/driver"
+)
+
+var (
+	srcDir         string
+	dbPrefix       string
+	dbPostfix      string
+	dbSeparator    string
+	restoreTimeout int
+)
+
+func amendDatabaseName(dbname string) string {
+	var tags []string
+
+	if dbPrefix != "" {
+		tags = append(tags, dbPrefix)
+	}
+
+	tags = append(tags, dbname)
+
+	if dbPostfix != "" {
+		tags = append(tags, dbPostfix)
+	}
+
+	return strings.Join(tags, dbSeparator)
+}
+
+func findDatabases() ([]string, error) {
+	databases := make([]string, 0)
+
+	entries, err := os.ReadDir(srcDir)
+	if err != nil {
+		return nil, util.Error(err, "error reading directory")
+	}
+
+	// See if db names need to be filtered
+	dbPattern := `.*`
+	if len(dbFilter) > 0 {
+		dbPattern = strings.Join(dbFilter, "|")
+	}
+
+	// Obtain database names from schema files
+	schemaPattern := fmt.Sprintf(
+		`^(?P<name>(%s))[.]%s$`,
+		dbPattern,
+		schemaFileExtension)
+	re := regexp.MustCompile(schemaPattern)
+
+	for _, v := range entries {
+		if v.IsDir() {
+			continue
+		}
+
+		// Extract database name from filename
+		match := re.FindStringSubmatch(v.Name())
+		if len(match) == 0 {
+			continue
+		}
+
+		databases = append(databases, match[1])
+	}
+
+	return databases, nil
+}
+
+func findCollections(db string) ([]string, error) {
+	collections := make([]string, 0)
+
+	entries, err := os.ReadDir(srcDir)
+	if err != nil {
+		return nil, util.Error(err, "error reading directory")
+	}
+
+	// See if collections need to be filtered
+	collPattern := `.*`
+	if len(collectionFilter) > 0 {
+		collPattern = strings.Join(collectionFilter, "|")
+	}
+
+	// Obtain schema names from json files
+	backupPattern := fmt.Sprintf(
+		`^%s[.](?P<collection>(%s))[.]%s$`,
+		db,
+		collPattern,
+		backupFileExtension)
+	re := regexp.MustCompile(backupPattern)
+
+	for _, v := range entries {
+		if v.IsDir() {
+			continue
+		}
+
+		// Extract database name from filename
+		match := re.FindStringSubmatch(v.Name())
+		if len(match) == 0 {
+			continue
+		}
+
+		collections = append(collections, match[1])
+	}
+
+	return collections, nil
+}
+
+func createDatabase(ctx context.Context, db string) {
+	withLogin(ctx, func(ctx context.Context) error {
```

Comment:
withLogin is not require here so as restoreDatabase is already wrapped in it.
Just use client.Get().CreateDatabase() in restoreDatabase directly.

Reply:
Thanks, got rid of it!